packages/react-charting/src/utilities/utilities.ts

```diff
@@ -1109,22 +1110,96 @@ export function wrapTextInsideDonut(selectorClass: string, maxWidth: number) {
       .attr('y', y)
       .attr('dy', lineNumber++ * lineHeight + 'em');
 
+    // Determine the length of a line
+    const lineLength = tspan.node()!.getBoundingClientRect().y;
+
+    // Determine the ellipsis length for word truncation
+    const ellipsis = '...';
+    tspan.text(ellipsis);
+    const ellipsisLength = tspan.node()!.getComputedTextLength();
+    tspan.text(null);
     while ((word = words.pop()!)) {
       line.push(word);
       tspan.text(line.join(' ') + ' ');
-      if (tspan.node()!.getComputedTextLength() > maxWidth && line.length > 1) {
+      // Determine if wrapping is required
+      // If yes, append a new line only if it does not exceed the max height
+      // here, max width = max height, so maxWidth is reused to check max height
+      if (
+        tspan.node()!.getComputedTextLength() > maxWidth &&
+        line.length > 1 &&
+        tspan.node()!.getBoundingClientRect().y < maxWidth
+      ) {
         line.pop();
         tspan.text(line.join(' ') + ' ');
-        line = [word];
-        tspan = text
-          .append('tspan')
-          .attr('id', `WordBreakId-${idx}-${lineNumber}`)
-          .attr('x', 0)
-          .attr('y', y)
-          .attr('dy', lineNumber++ * lineHeight + 'em')
-          .text(word);
+        // append a line only if the line is within the max height
+        // else truncate horizontally and break
+        if (tspan.node()!.getBoundingClientRect().y + lineLength < maxWidth) {
+          line = [word];
+          tspan = text
+            .append('tspan')
+            .attr('id', `WordBreakId-${idx}-${lineNumber}`)
+            .attr('x', 0)
+            .attr('y', y)
+            .attr('dy', lineNumber++ * lineHeight + 'em')
+            .text(word);
+        } else {
+          truncateHorizontally(tspan, line, maxWidth, ellipsisLength);
+          break;
+        }
+      }
+      // Determine if truncation is required vertically
+      // If yes, remove the last line, append ellipsis
+      // to the last word of the previous line (truncate horizontally) and break
+      if (tspan.node()!.getBoundingClientRect().y > maxWidth && line.length >= 1) {
```

Comment:
this should be max height?

Reply:
since maxHeight is the same as the maxWidth, I have reused maxWidth